utils/src/test/java/software/amazon/awssdk/utils/cache/lru/LruCacheTest.java

```diff
@@ -0,0 +1,372 @@
+     * finishes.
+     */
+    private Future<Value> performAsyncGet(LruCache<Key, Value> supplier, Key key) {
+        return executorService.submit(() -> supplier.get(key));
+    }
+
+    /**
+     * Asynchronously perform multiple "gets" on the provided supplier, returning the collection of futures to be completed when
+     * the "get" finishes.
+     */
+    private List<Future<?>> performAsyncGets(LruCache<Key, Value> supplier, int count) {
+        List<Future<?>> futures = new ArrayList<>();
+        for (int i = 0; i < count; ++i) {
+            futures.add(performAsyncGet(supplier, simpleTestKeys.get(count % 2)));
+        }
+        allExecutions.addAll(futures);
+        return futures;
+    }
+
+    /**
+     * Wait for the provided future to complete, failing the test if it does not.
+     */
+    private void waitFor(Future<?> future) {
+        invokeSafely(() -> future.get(10, TimeUnit.SECONDS));
+    }
+
+    /**
+     * Wait for all futures in the provided collection fo complete, failing the test if they do not all complete.
+     */
+    private void waitFor(Collection<Future<?>> futures) {
+        futures.forEach(this::waitFor);
+    }
+
+    /**
+     * Wait for all async gets ever created by this class to complete, failing the test if they do not all complete.
+     */
+    private void waitForAsyncGetsToFinish() {
+        waitFor(allExecutions);
+    }
+
+    private static final class Key {
+        private final Integer key;
+
+        private Key(Integer key) {
+            this.key = key;
+        }
+
+        static Key create(Integer key) {
+            return new Key(key);
+        }
+
+        Integer key() {
+            return key;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if ((o == null) || getClass() != o.getClass()) {
+                return false;
+            }
+            Key that = (Key) o;
+            return Objects.equals(key, that.key);
+        }
+
+        @Override
+        public int hashCode() {
+            return key != null? key.hashCode() : 0;
+        }
+    }
+
+    private static final class Value {
+        private final String param1;
+
+        Value(String param1) {
+            this.param1 = param1;
+        }
+
+        String param1() {
+            return param1;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if ((o == null) || getClass() != o.getClass()) {
+                return false;
+            }
+            Value that = (Value) o;
+            return Objects.equals(param1, that.param1);
+        }
+
+        @Override
+        public int hashCode() {
+            return param1 != null ? param1.hashCode() : 0;
+        }
+    }
```

Comment:
Any particular reason to not just use Integer and String directly instead of an explicit Key and Value type?

Reply:
Started out generically but it reads simpler if it's using concrete types, agreed